common/security/src/lib.rs

```diff
@@ -0,0 +1,73 @@
+/// Utility macro for writing secure arithmetic operations in order to avoid
+/// integer overflows.
+///
+/// # Examples
+///
+/// ```
+///# use crate::security::checked;
+/// let a: i64 = 1;
+/// let b: i64 = 2;
+/// let c: i64 = 3;
+///
+/// assert_eq!(checked!(a + b).unwrap(), 3);
+/// assert_eq!(checked!(a + b + c).unwrap(), 6);
+///
+/// // When doing multiple different operations, it's important to use parentheses in order
+/// // to guarantee the order of operation!
+/// assert_eq!(checked!(a + ((b - c) * c)).unwrap(), -2);
+///
+/// // When using numeric literals, the compiler might not be able to infer the type properly,
+/// // so if it complains, just add the type to the number.
+/// assert_eq!(checked!(10_u32 / 2_u32).unwrap(), 5);
+/// assert_eq!(checked!(10_u32 * 2_u32).unwrap(), 20);
+/// assert_eq!(checked!(10_u32 - 2_u32).unwrap(), 8);
+/// assert_eq!(checked!(10_u32 + 2_u32).unwrap(), 12);
+///
+/// assert_eq!(checked!(10_u32 / (1_u32 + 1_u32)).unwrap(), 5);
+/// assert_eq!(checked!(10_u32 * (1_u32 + 1_u32)).unwrap(), 20);
+/// assert_eq!(checked!(10_u32 - (1_u32 + 1_u32)).unwrap(), 8);
+/// assert_eq!(checked!(10_u32 + (1_u32 + 1_u32)).unwrap(), 12);
+///
+/// let max = u32::max_value();
+/// assert!(checked!(max + 1_u32).is_err());
+/// assert!(checked!(0_u32 - 1_u32).is_err());
+/// ```
+#[macro_export]
+macro_rules! checked {
+    ($a:tt + $b:tt) => {{
+        $a.checked_add($b).ok_or_else(|| format!("Operation results in overflow/underflow: {} + {}", $a, $b))
+    }};
+    ($a:tt - $b:tt) => {{
+        $a.checked_sub($b).ok_or_else(|| format!("Operation results in overflow/underflow: {} - {}", $a, $b))
+    }};
+    ($a:tt * $b:tt) => {{
+        $a.checked_mul($b).ok_or_else(|| format!("Operation results in overflow/underflow: {} * {}", $a, $b))
+    }};
+    ($a:tt / $b:tt) => {{
+        $a.checked_div($b).ok_or_else(|| format!("Operation results in overflow/underflow: {} / {}", $a, $b))
+    }};
+    ($a:tt + $($tokens:tt)*) => {{
+        match checked!( $($tokens)* ) {
```

Comment:
@xvschneider I think the other crate you found was transforming things into:

```
a.checked_add(b).and_then(|x| x.checked_add(c))
```

instead, which perhaps has less overhead? Not sure

Reply:
here it is: https://docs.rs/checked_ops/0.1.0/checked_ops/

Reply:
@mimoo sorry just catching up with this. This crate ^ is superior in that it honors operator precedence, I guess its not a big deal because we could enforce parentheses usage but could be an issue.

Reply:
is there a lint to enforce parentheses usage btw? I've always thought that we should enforce it via a lint to avoid stupid mistakes

Reply:
Even better than the lint rule, we could modify the macro to fail if there are no parentheses when more than 1 operation needs to be done